docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,1259 @@
+
+### Recovery
+
+#### Current Recovery Mechanisms
+
+Gardener and upstream Kubernetes already provide recovery mechanism for node and pod recovery in case of a failure of a node. Those have been tested in the scope of a [availability zone outage simulation](#availability-zone-outage-simulation).
+
+_Machine recovery_
+
+In the seed control plane, `Kube Controller Manager` will detect that a node has not renewed its lease and after a timeout (usually 40 seconds - configurable via `--node-monitor-grace-period`) it will transition the `Node` to `Unknown` state. `Machine-Controller-Manager` will detect that an existing `Node` has transitioned to `Unknown` state and will do the following:
+* It will transition the corresponding `Machine` to `Failed` state after waiting for a duration(currently 10 mins configured via [machine-health-timeout](https://github.com/gardener/gardener-extension-provider-gcp/blob/dff3d2417ff732fcce69ba10bbe5d04e13781539/charts/internal/machine-controller-manager/seed/templates/deployment.yaml)).
+* Thereafter a `deletion timestamp` will be put on this machine indicating that the machine is now going to be terminated, transitioning the machine to `Terminating` state.
+* It attempts to drain the node first and if it is unable to drain the node, then currently after a period of 2 hours (configurable via `machine-drain-timeout`), it will attempt to force-delete the `Machine` and create a new machine.
+
+In case `Machine-Controller-Manager` is unable to delete a machine, then that machine will be stuck in `Terminating` state. It will attempt to launch a new machine and if that also fails, then the new machine will transition to `CrashLoopBackoff` and will be stuck in this state.
+
+_Pod recovery_
+
+Once `Kube Controller Manager` transitions the node to `Unknown`/`NotReady`, it also puts the following taints on the node:
+```yaml
+taints:
+- effect: NoSchedule
+  key: node.kubernetes.io/unreachable
+- effect: NoExecute
+  key: node.kubernetes.io/unreachable
+```
+
+This annotation has the following effect:
+* New pods will not be scheduled unless they have a toleration added which is all permissive or matches the effect and/or key.
+
+In case where `kubelet` cannot reach the control plane, evicting the pod is not possible without intervention. Gardener provides [garbage collection](https://github.com/gardener/gardener/blob/4e2d28f2af7093eb94819652a6f4709b5fbfaf06/pkg/gardenlet/controller/shoot/shoot_care_control.go#L343) which [forcefully deletes](https://github.com/gardener/gardener/blob/4e2d28f2af7093eb94819652a6f4709b5fbfaf06/pkg/operation/care/garbage_collection.go#L117-L138) pods after currently configured duration of 5 mins. This will force the Kube Scheduler to reschedule these pods to nodes in zones that are available.
```

Comment:
The point is more that Kubelet is not able to talk to the Kube-Apiserver to confirm the pod deletion. See https://kubernetes.io/docs/concepts/architecture/nodes/#condition:

> In some cases when the node is unreachable, the API server is unable to communicate with the kubelet on the node. The decision to delete the pods cannot be communicated to the kubelet until communication with the API server is re-established. In the meantime, the pods that are scheduled for deletion may continue to run on the partitioned node.

> The node controller does not force delete pods until it is confirmed that they have stopped running in the cluster. You can see the pods that might be running on an unreachable node as being in the Terminating or Unknown state. In cases where Kubernetes cannot deduce from the underlying infrastructure if a node has permanently left a cluster, the cluster administrator may need to delete the node object by hand.

The garbage collector also deletes `Terminating` pods which helps to recover faster. However, looking at the code again it only does it for the `Kube-System` namespace. @unmarshall I think we have a gap here.

Reply:
> The garbage collector also deletes `Terminating` pods which helps to recover faster. However, looking at the code again it only does it for the `Kube-System` namespace

Good point, I missed this handling! You are right that it deletes `Terminating` (for over 5mins) Pods in the kube-system namespace of the Shoots. But when I check https://github.com/gardener/gardener/pull/383 (the PR that introduced it) I don't think it is a recovery option (it is only for Pods in the kube-system namespace) but rather a mechanism introduced to reduce the noise for operators.

Reply:
>I guess the recovery option is rather that if kubelet is unhealthy, MCM will remove the Machine after the health period passes and new Pods will be scheduled on healthy Nodes.

We simulated a case where we disabled all ingress/egress to/from a zone. We also changed `cloudprovider` credentials so that MCM is unable to successfully invoke provider APIs to manage machines. Even in this case we saw that all terminating pods which had relaxed anti-affinity policies were scheduled onto other nodes in another zone. The old machines were stuck in `Unknown` state and when MCM tried to start new machines then those newer machines are now stuck in `Terminating` state. So in this case MCM will not be able to recover any machines.

Reply:
Ok there was an error in our test. The `cloudprovider` secret was getting reset back to the correct value after 5mins. This time i changed also the `seed-provider-secret` in the project namespace (virtual garden cluster) and also changed `cloudprovider` secret to have the same secret. After that i ran the test and observed that the machines are stuck in `Terminating` state and the new ones are going from `Pending -> CrashLoopBackOff`. Pods are stuck in terminating state as well:
```
```

Reply:
I have changed this section. Please have a look now.